controllers/agentcluster_controller.go

```diff
@@ -155,6 +221,52 @@ func (r *AgentClusterReconciler) createClusterDeployment(ctx context.Context, lo
 			},
 		},
 	}
+
+	var ok bool
+	clusterDeployment.Spec.BaseDomain, ok, err = unstructured.NestedString(controlPlane.UnstructuredContent(), "spec", "dns", "baseDomain")
+	if !ok {
+		log.Infof("Control plane object has no baseDomain, waiting more")
+		return nil, nil
+	}
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get base domain")
+	}
+
+	pullSecretName, ok, err := unstructured.NestedString(controlPlane.UnstructuredContent(), "spec", "pullSecret", "name")
+	if !ok {
+		log.Infof("Control plane object has no pullSecretName, waiting more")
+		return nil, nil
+	}
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get pull secret ref")
+	}
+	clusterDeployment.Spec.PullSecretRef = &corev1.LocalObjectReference{Name: pullSecretName}
+
+	return clusterDeployment, nil
+}
+
+func (r *AgentClusterReconciler) createClusterDeployment(ctx context.Context, log logrus.FieldLogger, agentCluster *capiproviderv1alpha1.AgentCluster) (ctrl.Result, error) {
+	controlPlane, err := r.getUnstructuredControlPlane(ctx, log, agentCluster)
+	if controlPlane == nil {
+		result := ctrl.Result{}
+		if err != nil {
+			result.Requeue = true
+		}
+		return result, err
+	}
+
+	// Ensure we add a watcher to the control plane object.
+	err = r.addControlPlaneWatcher(ctx, log, controlPlane)
```

Comment:
Why do we need a watcher here?
We call this method once!

Reply:
Also, do we think the parameters we get from the control plane might change?

Reply:
@avishayt what do you think?